app/backend/src/tests/test_auth.py

```diff
@@ -309,4 +309,272 @@ def test_signup_invalid_birthdate(db):
             assert e.value.details() == errors.INVALID_BIRTHDATE
 
 
+def test_invalid_signup_token_info(db):
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        with pytest.raises(grpc.RpcError) as e:
+            reply = auth_api.SignupTokenInfo(auth_pb2.SignupTokenInfoReq(signup_token="notarealtoken"))
+        assert e.value.code() == grpc.StatusCode.NOT_FOUND
+        assert e.value.details() == errors.INVALID_TOKEN
+
+
+def test_signup_invalid_email(db):
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        reply = auth_api.Signup(auth_pb2.SignupReq(email="a"))
+    assert reply.next_step == auth_pb2.SignupRes.SignupStep.INVALID_EMAIL
+
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        reply = auth_api.Signup(auth_pb2.SignupReq(email="a@b"))
+    assert reply.next_step == auth_pb2.SignupRes.SignupStep.INVALID_EMAIL
+
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        reply = auth_api.Signup(auth_pb2.SignupReq(email="a@b."))
+    assert reply.next_step == auth_pb2.SignupRes.SignupStep.INVALID_EMAIL
+
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        reply = auth_api.Signup(auth_pb2.SignupReq(email="a@b.c"))
+    assert reply.next_step == auth_pb2.SignupRes.SignupStep.INVALID_EMAIL
+
+
+def test_signup_existing_email(db):
+    # Signed up user
+    user, _ = generate_user()
+
+    # Attempt to signup again with the same email
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        reply = auth_api.Signup(auth_pb2.SignupReq(email=user.email))
+    assert reply.next_step == auth_pb2.SignupRes.SignupStep.EMAIL_EXISTS
+
+
+def test_successful_authenticate(db):
```

Comment:
Use the `fast_passwords` fixture in tests that do any password hashing: that will significantly speed the tests up.